pkg/apis/core/validation/validation.go

```diff
@@ -5360,9 +5360,33 @@ func ValidateSecurityContext(sc *core.SecurityContext, fldPath *field.Path) fiel
 // maxGMSACredentialSpecLength is the max length, in bytes, for the actual contents
 // of a GMSA cred spec. In general, those shouldn't be more than a few hundred bytes,
 // so we want to give plenty of room here while still providing an upper bound.
+// The runAsUserName field will be used to execute the given container's entrypoint, and
+// it can be formatted as "DOMAIN/USER", where the DOMAIN is optional, maxRunAsUserNameDomainLength
+// is the max character length for the user's DOMAIN, and maxRunAsUserNameUserLength
+// is the max character length for the USER itself. Both the DOMAIN and USER have their
+// own restrictions, and more information about them can be found here:
+// https://support.microsoft.com/en-us/help/909264/naming-conventions-in-active-directory-for-computers-domains-sites-and
+//  https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.localaccounts/new-localuser?view=powershell-5.1
 const (
 	maxGMSACredentialSpecLengthInKiB = 64
 	maxGMSACredentialSpecLength      = maxGMSACredentialSpecLengthInKiB * 1024
+	maxRunAsUserNameDomainLength     = 256
+	maxRunAsUserNameUserLength       = 21
+)
+
+var (
+	// a valid NetBios Domain name cannot start with a dot, has at least 1 character,
+	// at most 15 characters, and it cannot the characters: \ / : * ? " < > |
+	// we're also excluding the control characters (ASCII codes: 0-31, 127; in hexa: 00-1F, 7F)
+	validNetBiosRegex, _ = regexp.Compile(`^(?:[^\x00-\x1F\x7F\\/:\*\?"<>|\.][^\x00-\x1F\x7F\\/:\*\?"<>|]{0,14})?$`)
+
+	// a valid DNS name contains only alphanumeric characters, dots, and dashes.
+	validDNSRegex, _ = regexp.Compile(`^(?:(?:[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?\.)+[a-zA-Z0-9][a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?$`)
+
+	// a username is invalid if it contains the characters: " / \ [ ] : ; | = , + * ? < > @
+	// or it contains only dots or spaces.
+	// control characters will also be considered invalid (ASCII codes: 0-31, 127; in hexa: 00-1F, 7F)
+	invalidUserNameRegex, _ = regexp.Compile(`[\x00-\x1F\x7F"/\\:;|=,\+\*\?<>@\[\]]|^[\. ]{1,20}$`)
```

Comment:
is there a reason `^[\. ]{1,20}$` limited to 20 characters and didn't do `^[\. ]+$`? making this regex know about the max username length doesn't seem necessary.

joining the validation in a single invalid regex makes it hard to give good error messages... ideally a control character would give one error message, an invalid character would give another (including the list of invalid characters), and all periods/spaces would give another.

Reply:
regexes are a very compact way to represent validation rules, but are also error-prone once they get reasonable complex. for the domain validation, it would be easier to follow code that does each of these as distinct steps
* checks length
* checks valid leading character
* checks absence of control characters
* checks absence of invalid printable characters